### ISSUE TITLE
Issue with using a different evaluation grid

### ISSUE THREAD

**Issue by Katarina-Poole** [Any2HRTF/Mesh2HRTF#78]
I've been trying to run the mesh2HRTF with a different evaluation grid but unfortunately it seems to fail. The error it throws is:

`Traceback (most recent call last):
  File "/rds/general/user/kp1912/home/GitHubRepos/SONICOM/toolboxes/mesh2hrtf-tools/manage_numcalc_script.py", line 106, in <module>
    m2h.manage_numcalc(
  File "/rds/general/user/kp1912/home/GitHubRepos/SONICOM/toolboxes/Mesh2HRTF-develop/mesh2hrtf/NumCalc/manage_numcalc.py", line 226, in manage_numcalc
    all_instances, instances_to_run, *_ = _check_project(
  File "/rds/general/user/kp1912/home/GitHubRepos/SONICOM/toolboxes/Mesh2HRTF-develop/mesh2hrtf/NumCalc/manage_numcalc.py", line 528, in _check_project
    estimates = np.concatenate(
  File "<__array_function__ internals>", line 180, in concatenate
ValueError: all the input arrays must have same number of dimensions, but the array at index 0 has 2 dimension(s) and the array at index 1 has 1 dimension(s`

With a bit more digging it seems for some reason NumCalc fails to run estimates (NumCalc -estimate_ram) when this new evaluation grid is used, so generates an empty Memory.txt file (which then gives an empty estimates array leaving to the above error).

Not sure why this is the case? Using the Default and ARI evaluation grids works fine and the evaluation grid I generated with the write-evaluation grid script looks normal. 

Any advice would be more than welcome! I've attached the elements and nodes txt files for the evaluation grid I'm trying to use
[Elements.txt](https://github.com/Any2HRTF/Mesh2HRTF/files/10754851/Elements.txt)
[Nodes.txt](https://github.com/Any2HRTF/Mesh2HRTF/files/10754853/Nodes.txt)

**Comment by SDX-LV** [comment]
Hi, you have almost the same error as in the issue https://github.com/Any2HRTF/Mesh2HRTF/issues/72
Can it be that your evaluation grid also contains multiple surfaces instead of one big shell?

Also you can generate plots and human-readable node coordinates if you put your evaluation grid definition into this script: https://sourceforge.net/p/mesh2hrtf-tools/code/ci/master/tree/extra_Evaluation_Grids/generate_mesh2hrtf_evaluation_grid.py

**Comment by terrainer** [comment]
> 

The issue is almost definitely the same as #72, and the evaluation grids provided in that link also produce this error.

**Comment by f-brinkmann** [comment]
I quickly wrote a check for isolated nodes that we could include in the Blender export plug-in. But your mesh seems to be fine, which is also suggested by the additional plot. Could you run `numcalc -estimate_ram` in isolation to see if that gives any output (also check or upload the output files for hints).

@terrainer, @SDX-LV  can you check if the isolated nodes in your grid multi-radius grid is detected?

```python
import numpy as np
import os
import mesh2hrtf as m2h


def check_evaluation_grid(name):
    """
    Check if an evaluation grid contains an isolated node, i.e., a node that
    is not contained in any element. Mesh2HRTF can not handle this.

    Parameters
    ----------
    name : str
        Name of the folder containing the nodes and elements of the evaluation
        grid in Nodes.txt and Elements.txt

    Returns
    -------
    isolated_node : bool
        ``True`` if an isolated node was detected, ``False`` otherwise.
    """

    # check if the grid exists
    if not os.path.isfile(os.path.join(name, "Nodes.txt")):
        raise ValueError(f"{os.path.join(name, 'Nodes.txt')} does not exist")

    # read the nodes ----------------------------------------------------------
    with open(os.path.join(name, "Nodes.txt"), "r") as f_id:
        nodes = f_id.readlines()

    # get number of nodes
    N_nodes = int(nodes[0].strip())
    node_ids = np.zeros(N_nodes, dtype=int)

    # get node ids
    for nn in range(N_nodes):
        node = nodes[nn+1].strip().split(" ")
        node_ids[nn] = int(node[0])

    # get elements ------------------------------------------------------------
    with open(os.path.join(name, "Elements.txt"), "r") as f_id:
        elements = f_id.readlines()

    # get number of elements and element vertices
    N_elements = int(elements[0].strip())
    M_vertices = len(elements[1].strip().split(" ")) - 4
    element_array = np.zeros((N_elements, M_vertices), dtype=int)

    # get node ids
    for nn in range(N_elements):
        element = elements[nn+1].strip().split(" ")
        for mm in range(M_vertices):
            element_array[nn, mm] = int(element[mm+1])

    # check for isolated node -------------------------------------------------
    # (node that is not contained in any element)
    element_array = np.unique(element_array.flatten())

    isolated_node = False
    for node_id in node_ids:
        if node_id not in element_array:
            isolated_node = True
            break

    return isolated_node

# check grid
name = r"Grid"
isolated_node = check_evaluation_grid(name)

# plot gird
m2h.read_evaluation_grid(name, True)
```

**Comment by f-brinkmann** [comment]
another thing that came to my mind what BEM-Method are you using. @schaastrommla if I remember it correctly estimate ram is only implemented for FMM-ML-BEM - is that right?

**Comment by Katarina-Poole** [comment]
Thanks for the speedy replies! So I generated the grid based on measured positions recorded from a sofa file. 
I did realise there are probably non-unique values in the grid (i.e. exactly above the listener - where there where multiple measurements) which I removed and regenerated the grid but still ran into the same error 

Here it is plotted: 
![image](https://user-images.githubusercontent.com/58176031/219615100-bd358a80-e6ec-4743-a19d-2bd3be91b67c.png)

Here it is as a csv in polar coordinates (index; azimuth; elevation; distance) for easy reading (they were converted to Cartesian for write_evaluation_grid).
[SONICOMpolarcoords.csv](https://github.com/Any2HRTF/Mesh2HRTF/files/10765639/SONICOMpolarcoords.csv)

And updated nodes and elements
[Nodes.txt](https://github.com/Any2HRTF/Mesh2HRTF/files/10765628/Nodes.txt)
[Elements.txt](https://github.com/Any2HRTF/Mesh2HRTF/files/10765630/Elements.txt)

The BEM method I'm using is ML-FMM BEM (I just stuck with the defaults for now from the blender export script - apart from the default eval grids)

However looking at NC.out, it does look like its found an isolated node which is probably whats causing this error like in #74.
 
Start time: 17/2/2023 10:34:6

>> G E N E R A L   P A R A M E T E R S <<

3D analysis

number of the isolated node: 200000
Isolated node found!

How would I get rid of the isolated node? Looks like its the first entry in the nodes.txt file but not sure what coordinate that would relate to

**Comment by schaastrommla** [comment]
@Katarina-Poole  the error message is a bit strange, because according to your Elements.txt  the node 200000 is for example in element 200315, so 200000 should not be isolated, is the file path to Element.txt correct in NC.inp? Or maybe the error message is outdated?

**Comment by Katarina-Poole** [comment]
@schaastrommla  I'm pretty sure the most recent nodes and elements files I attached are from the project I just generated the isolated node error. I've reattached here again
[Elements.txt](https://github.com/Any2HRTF/Mesh2HRTF/files/10766240/Elements.txt)
[Nodes.txt](https://github.com/Any2HRTF/Mesh2HRTF/files/10766241/Nodes.txt)

I do also have it run the default evaluation grid in the same simulation run. Its nodes and elements are here:
[Nodes.txt](https://github.com/Any2HRTF/Mesh2HRTF/files/10766244/Nodes.txt)
[Elements.txt](https://github.com/Any2HRTF/Mesh2HRTF/files/10766246/Elements.txt)

But I assumed it would be fine for the default as it has been before

**Comment by Katarina-Poole** [comment]
FYI this is the log file output:
```
Starting manage_numcalc with the following arguments [Feb 17 2023, 10:34:05]
----------------------------------------------------------------------------
project_path: /rds/general/user/kp1912/home/GitHubRepos/SONICOM/toolboxes/mesh2hrtf-tools
numcalc_path: /rds/general/user/kp1912/home/GitHubRepos/SONICOM/toolboxes/Mesh2HRTF-develop/mesh2hrtf/NumCalc/bin/NumCalc
max_ram_load: 128.0
ram_safety_factor: 1.05
max_cpu_load: 90
max_instances: 8
wait_time: 15
starting_order: alternate
confirm_errors: False

NumCalc executable: /rds/general/user/kp1912/home/GitHubRepos/SONICOM/toolboxes/Mesh2HRTF-develop/mesh2hrtf/NumCalc/bin/NumCalc

Obtaining RAM estimates for /rds/general/user/kp1912/home/GitHubRepos/SONICOM/toolboxes/mesh2hrtf-tools/myHRTF_project_R/NumCalc/source_1
```

This is the error output (after running it once where it fails to produce the estimates):

```
Traceback (most recent call last):
  File "/rds/general/user/kp1912/home/GitHubRepos/SONICOM/toolboxes/mesh2hrtf-tools/manage_numcalc_script.py", line 106, in <module>
    m2h.manage_numcalc(
  File "/rds/general/user/kp1912/home/GitHubRepos/SONICOM/toolboxes/Mesh2HRTF-develop/mesh2hrtf/NumCalc/manage_numcalc.py", line 226, in manage_numcalc
    all_instances, instances_to_run, *_ = _check_project(
  File "/rds/general/user/kp1912/home/GitHubRepos/SONICOM/toolboxes/Mesh2HRTF-develop/mesh2hrtf/NumCalc/manage_numcalc.py", line 516, in _check_project
    subprocess.run(
  File "/rds/general/user/kp1912/home/anaconda3/envs/meshHRTF/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/rds/general/user/kp1912/home/GitHubRepos/SONICOM/toolboxes/Mesh2HRTF-develop/mesh2hrtf/NumCalc/bin/NumCalc -estimate_ram']' returned non-zero exit status 1.
```

**Comment by terrainer** [comment]
> @Katarina-Poole the error message is a bit strange, because according to your Elements.txt the node 200000 is for example in element 200315, so 200000 should not be isolated, is the file path to Element.txt correct in NC.inp? Or maybe the error message is outdated?

Not outdated, as this error occured for me on a fresh project with a fresh install of Mesh2HRTF and Mesh2HRTF-tools.

> @f-brinkmann I quickly wrote a check for isolated nodes that we could include in the Blender export plug-in. But your mesh seems to be fine, which is also suggested by the additional plot. Could you run NumCalc -estimate_ram in isolation to see if that gives any output (also check or upload the output files for hints).

@terrainer, @SDX-LV can you check if the isolated nodes in your grid multi-radius grid is detected?

Not sure how to run this, but I'm happy to try it with some instruction.

**Comment by SDX-LV** [comment]
@terrainer - let's keep the issues separate. I think this issue will be soon resolved, while for multi-sphere simulations the solution also exists. There each radius can be simulated as a separate evaluation grid. We are only missing a script to merge multiple resulting SOFA files into a single multi-field SOFA. But it is easier to make such post-processing script than to re-develop Mesh2HRTF numCalc solver to support multiple radiuses in the same evaluation grid.

**Comment by terrainer** [comment]
> @terrainer - let's keep the issues separate. I think this issue will be soon resolved, while for multi-sphere simulations the solution also exists. There each radius can be simulated as a separate evaluation grid. We are only missing a script to merge multiple resulting SOFA files into a single multi-field SOFA. But it is easier to make such post-processing script than to re-develop Mesh2HRTF numCalc solver to support multiple radiuses in the same evaluation grid.

This is the same issue - I am using the separated evaluation grids and the same problem occurs.

**Comment by f-brinkmann** [comment]
Hi everyone - sorry for the silence. The end of the semester was a bit work heavy on my side.

@Katarina-Poole I noted that both your grids start with the same ID (the numbers in the first column of your Node and Element files. Both grids start with 200000). Can you try two things please: Test if the simulation works, if you only use one grid. Test if the simulation works with both grids, after modifing the custom grid to start at 300000. This can be done with https://mesh2hrtf.readthedocs.io/en/latest/mesh2hrtf.html#mesh2hrtf.write_evaluation_grid

@terrainer might that also be your problem? If yes, it would be appreciated if you ran the same tests. If you upload your multi-radius grid (the one containing all radii in one gird) I could check if my test for isolated nodes works.

If this would be the problem, it would be straight forward to catch upon export from Blender.

**Comment by petibub** [comment]
Hi! 
As for different radii in the evaluation grid for Mesh2HRTF: Mesh2HRTF should be able to do that, but let me re-check our acoustician. 
As for different radii in a SOFA file: SOFA does NOT link the source positions by radius or any other dimension. Each source position is given by an independent 3D vector, and thus each source position is independent from others, and these positions can be arbitrary. Examples of handling SOFA files with different radii can be found in these demos: [demo_SOFAmerge](https://github.com/sofacoustics/SOFAtoolbox/blob/master/SOFAtoolbox/demos/demo_SOFAmerge.m) and [demo_SCUT2SOFA](https://github.com/sofacoustics/SOFAtoolbox/blob/master/SOFAtoolbox/demos/demo_SCUT2SOFA.m).

**Comment by terrainer** [comment]
> Hi everyone - sorry for the silence. The end of the semester was a bit work heavy on my side.
> 
> @Katarina-Poole I noted that both your grids start with the same ID (the numbers in the first column of your Node and Element files. Both grids start with 200000). Can you try two things please: Test if the simulation works, if you only use one grid. Test if the simulation works with both grids, after modifing the custom grid to start at 300000. This can be done with https://mesh2hrtf.readthedocs.io/en/latest/mesh2hrtf.html#mesh2hrtf.write_evaluation_grid
> 
> @terrainer might that also be your problem? If yes, it would be appreciated if you ran the same tests. If you upload your multi-radius grid (the one containing all radii in one gird) I could check if my test for isolated nodes works.
> 
> If this would be the problem, it would be straight forward to catch upon export from Blender.

It appears that changing the elements and node numbers has worked. It would be incredibly useful to have the Blender export be able to automatically handle this.

**Comment by f-brinkmann** [comment]
agreed. I'll open a separate issue for this ans will close this one after final confirmation from Katarina. @terrainer can you upload your multi-radius grid or does that now work as well?

**Comment by terrainer** [comment]
> agreed. I'll open a separate issue for this ans will close this one after final confirmation from Katarina. @terrainer can you upload your multi-radius grid or does that now work as well?

Multi-radius seems to work, however I'm now having a problem with finalizing, so I'll open another issue.

**Comment by terrainer** [comment]
The multi radius grid works when split and with the elements renumbered.

**Comment by SDX-LV** [comment]
> Hi! As for different radii in the evaluation grid for Mesh2HRTF: Mesh2HRTF should be able to do that, but let me re-check our acoustician. As for different radii in a SOFA file: SOFA does NOT link the source positions by radius or any other dimension. Each source position is given by an independent 3D vector, and thus each source position is independent from others, and these positions can be arbitrary. Examples of handling SOFA files with different radii can be found in these demos: [demo_SOFAmerge](https://github.com/sofacoustics/SOFAtoolbox/blob/master/SOFAtoolbox/demos/demo_SOFAmerge.m) and [demo_SCUT2SOFA](https://github.com/sofacoustics/SOFAtoolbox/blob/master/SOFAtoolbox/demos/demo_SCUT2SOFA.m).

So there is actually a merging script in the Matlab SOFA toolbox already! Good to know :)

**Comment by SDX-LV** [comment]
> > Hi everyone - sorry for the silence. The end of the semester was a bit work heavy on my side.
> > @Katarina-Poole I noted that both your grids start with the same ID (the numbers in the first column of your Node and Element files. Both grids start with 200000). Can you try two things please: Test if the simulation works, if you only use one grid. Test if the simulation works with both grids, after modifing the custom grid to start at 300000. This can be done with https://mesh2hrtf.readthedocs.io/en/latest/mesh2hrtf.html#mesh2hrtf.write_evaluation_grid
> > @terrainer might that also be your problem? If yes, it would be appreciated if you ran the same tests. If you upload your multi-radius grid (the one containing all radii in one gird) I could check if my test for isolated nodes works.
> > If this would be the problem, it would be straight forward to catch upon export from Blender.
> 
> It appears that changing the elements and node numbers has worked. It would be incredibly useful to have the Blender export be able to automatically handle this.

I just got this...  So it currently is necessary that every evaluation grid that is included in the simulation has each node ID different from any other. my https://sourceforge.net/p/mesh2hrtf-tools/code/ci/master/tree/extra_Evaluation_Grids/generate_mesh2hrtf_evaluation_grid.py script did not handle this requirement. Need to think about a good solution.

**Comment by Katarina-Poole** [comment]
Sorry for the delay in the update, took me a while to run the simulation (the high performance computing cluster struggled for some reason, not sure if its due to HPC itself or something with the code (even though I was allocating large amounts of RAM, the output file would sometimes say 0 RAM available and then stop the code).

So finally ran mesh2hrtf on the SONICOM evaluation grid and it doesn't look that normal (here are some of the plots):

![HRIR_SONICOM_48000](https://user-images.githubusercontent.com/58176031/221526539-bd9d7aa8-b645-4d99-940c-e82febf75707.png)

![HRIR_SONICOM_48000_3D](https://user-images.githubusercontent.com/58176031/221526652-f77597ef-25b2-466d-8659-b39da9a9d3b5.jpeg)

![HRTF_SONICOM_48000_3D](https://user-images.githubusercontent.com/58176031/221526682-f9a0bffc-800c-4c65-8ef1-eca41f2b3c0b.jpeg)


And then a plot of one of the the locations (measured and modelled for comparison). Can see some wierd ringing at the beginning of the impulse response for the  simulated HRTF

![image](https://user-images.githubusercontent.com/58176031/221527262-d61445d8-7254-4e7d-be90-222687196543.png)

I'm going to do some more digging as to why, but if you have any ideas why this might be the case, please let me know :)

**Comment by f-brinkmann** [comment]
I think one problem is that the HRIRs are not plotted in the intended order. This bug will be fixed with the next pyfar release (https://github.com/pyfar/pyfar/issues/383). For now you have to manually sort the data for plotting.

**Comment by Katarina-Poole** [comment]
Yup, looks sensible with sorting
![image](https://user-images.githubusercontent.com/58176031/221536873-811b70fe-b2a5-4a30-9add-7e353bb15d97.png)

**Comment by f-brinkmann** [comment]
closed by #86

**Comment by mberz** [comment]
PR https://github.com/pyfar/pyfar/issues/383 is merged, the issue is fixed in pyfar versions >= 0.5.3
